system/jlib/jptree.ipp

```diff
@@ -404,14 +338,240 @@ private:
+{
+    static inline unsigned getHash(const char *k)
+    {
+        return hashnc((const byte *)k,strlen(k),17);
+    }
+
+    inline bool eq(const char *k)
+    {
+        return stricmp(k,str)==0;
+    }
+
+
+    static AttrStrNC *create(const char *k)
+    {
+        size32_t kl = (k?strlen(k):0);
+        AttrStrNC *ret = (AttrStrNC *)malloc(sizeof(AttrStrNC)+kl);
+        memcpy(ret->str,k,kl);
+        ret->str[kl] = 0;
+        ret->hash = hashnc((const byte *)k,kl,17);
+        ret->linkcount = 0;
+        return ret;
+    }
+
+    static void destroy(AttrStrNC *a)
+    {
+        free(a);
+    }
+
+};
+
+class CAttrValHashTable
+{
+    CMinHashTable<AttrStrC>  htc;
+    CMinHashTable<AttrStrNC> htnc;
+    CMinHashTable<AttrStrC>  htv;
+
+public:
+    inline AttrStr *addkey(const char *v,bool nc)
+    {
+
+        AttrStr * ret;
+        if (nc)
+            ret = htnc.find(v,true);
+        else
+            ret = htc.find(v,true);
+        if (ret->linkcount!=(unsigned short)-1)
+            ret->linkcount++;
+        return ret;
+    }
+
+    inline AttrStr *addval(const char *v)
+    {
+
+        AttrStr * ret = htv.find(v,true);
+        if (ret->linkcount!=(unsigned short)-1)
+            ret->linkcount++;
+        return ret;
+    }
+
+    inline void removekey(AttrStr *a,bool nc)
+    {
+        if (a->linkcount!=(unsigned short)-1)
+        {
+            if (--(a->linkcount)==0)
+            {
+                if (nc)
+                    htnc.remove((AttrStrNC *)a);
+                else
+                    htc.remove((AttrStrC *)a);
+            }
+        }
+    }
+
+    inline void removeval(AttrStr *a)
+    {
+        if (a->linkcount!=(unsigned short)-1)
+            if (--(a->linkcount)==0)
+                htv.remove((AttrStrC *)a);
+    }
+
+};
+
+class jlib_decl AtomChildMap : public ChildMap
+{
+public:
+// SuperHashTable definitions
+    virtual unsigned getHashFromElement(const void *e) const override;
+
+    virtual unsigned getHashFromFindParam(const void *fp) const override
+    {
+        return hashc((const unsigned char *)fp, (size32_t)strlen((const char *)fp), 0);
+    }
+    virtual bool matchesFindParam(const void *e, const void *fp, unsigned fphash) const override
+    {
+        return (0 == strcmp(((IPropertyTree *)e)->queryName(), (const char *)fp));
+    }
+};
+
+// case insensitive childmap
+class jlib_decl AtomChildMapNC : public ChildMap
+{
+public:
+// SuperHashTable definitions
+    virtual unsigned getHashFromElement(const void *e) const override;
+    virtual unsigned getHashFromFindParam(const void *fp) const override
+    {
+        return hashnc((const unsigned char *)fp, (size32_t)strlen((const char *)fp), 0);
+    }
+    virtual bool matchesFindParam(const void *e, const void *fp, unsigned fphash) const override
+    {
+        return (0 == stricmp(((IPropertyTree *)e)->queryName(), (const char *)fp));
+    }
+};
+
+
+
+class jlib_decl CAtomPTree : public PTree
```

Comment:
I seem to have lost a comment here as well.

I'm not keen on the duplication of the structure.  E.g., having two completely different ways of allocating attribute arrays.  It would be simpler to always create AttrStrs and store those in the lists - the difference being that one was commoned up, the other left as it is?